### DISCUSSION TITLE
Bug or configuration error? (PHP Version used)

### DISCUSSION THREAD

**Discussion by internalsystemerror** [renovatebot/renovate#13637] [Request Help] [answered]
Hi, I'm investigating implementing renovate for Laminas framework (https://getlaminas.org/).

The configuration I'm testing is here: https://github.com/laminas/.github/pull/19.

There is an issue when updating the lockfile and the problem is one mentioned in a previous discussion. The highest available PHP version is always used. For example a constraint of `~7.4 || ~8.0.0 || ~8.1.0` will use an image with version `8.1.1`.

The checks run using each of those versions and since packages are installed that require PHP 8, our checks for PHP 7.4 fail and we would be unable to get any lock file updates. I would suggest that the lowest minor PHP version is used as PHP 7 code will for the vast majority of cases, work on PHP 8. However the reverse is not true.

That's not a rule that applies to everyone though, so a configuration option here might be preferable, highest vs lowest language version.

An example of this issue occurring can be found here: https://github.com/internalsystemerror/laminas-servicemanager/pull/4.

One possible workaround that I've not tried is to manually specify the PHP version in the config section of each composer.json file. There are a lot of repositories and this would then have other implications so I would rather avoid this at all costs.

**Comment by viceice** [comment, marked as answer]
Add `constraints.php="~7.4"` to renovate config.

https://docs.renovatebot.com/configuration-options/#constraints

**Reply by internalsystemerror** [reply]
Thank you but this would still require manually specifying the lower version. This would then need to be done for each individual repository as they don't all have the same version constraints.

If we do decide to do this, once done, maintaining it could be easy enough providing we don't forget. However, I was hoping more to be able to say choose the lower of the range rather than the higher?

**Reply by internalsystemerror** [reply]
The link you provided does not list php, however inspecting the code it looks like this should still work. I'll give it a try.

**Reply by internalsystemerror** [reply]
This does allow us to run the lockfile updates against a fixed PHP version, thank you.

**Reply by rarkins** [reply]
Can you fix the docs in your PR too?

**Reply by internalsystemerror** [reply]
The docs?

**Reply by rarkins** [reply]
For "constraints"

**Reply by internalsystemerror** [reply]
To which PR are you referring? If you mean the renovate configuration for laminas, I'm unaware of where this is incorrect? If you mean the one for yourselves that's not involving constraints?

**Reply by rarkins** [reply]
It's related to this:

> The link you provided does not list php

Although I just realized that it's supported *managers*, not languages, and composer is listed there. So you can disregard this suggestion

**Reply by internalsystemerror** [reply]
Ohhhh, I get you now sorry. I have tested a constraint of `php` and that appears to work as expected though.

**Comment by rarkins** [comment]
I'm not a regular PHP/Composer user, so I can't tell from your example immediately what's wrong. Do you mean that Renovate is implicitly "upgrading" the PHP version within that changed lockfile? e.g. it could be compared to going from npm@6 to npm@7 where the lock file format changed?

And if so, is there some obvious field or other way in which Renovate could automatically detect the version of PHP in use for the existing lockfile, in order to reuse the same?

**Reply by internalsystemerror** [reply]
What renovate is doing is correct, in that it's reading the `"php": "^7.4 || ~8.0.0 || ~8.1.0"` requirement from `composer.json` and it's determining that the docker image for PHP 8.1.1 should be used when performing the update.

However, this results in packages being updated which themselves only support PHP > 8.0. Meaning that the tests running against 7.4 are now failing since the lock file now includes packages which require PHP >= 8.0. You can see this here: https://github.com/internalsystemerror/laminas-servicemanager/actions/runs/1715906658

What we would prefer is to be able to tell renovate to use the lower bound of that PHP range (i.e. an image using PHP 7.4 should be used) instead of defaulting to the highest in the range.

**Reply by internalsystemerror** [reply]
Take the following `composer.json` file https://github.com/internalsystemerror/laminas-servicemanager/blob/3.11.x/composer.json. This has the constraint:

```json
 "require": {
        "php": "~7.4.0 || ~8.0.0 || ~8.1.0",
        ...
    },
```

Renovate detects this and loads a docker image of PHP 8.1.1. The lockfile is then updated using PHP 8.1.1 as the detected version and updates a transient dependency `psr/cache` from v1 to v3.

`psr/cache` v1 runs on PHP 7.4, however v3 requires a minimum of PHP 8.0. So when the lockfile is installed on PHP 7.4 you get the errors in the action linked above.

**Reply by internalsystemerror** [reply]
The option that @viceice suggested would possible fix this allowing us to specify the version we wish to use for generating the lockfile. However this differs from repository to repository depending on the packages we need and the minimum PHP version they support, along with which PHP versions still receive security updates.

So what we would like to do is to say use the lowest-minor version from that PHP range so that it can be applied across all 3 orgs.

**Reply by internalsystemerror** [reply]
One last thing just regarding whether its possible to autodetect the PHP version from the lockfile. Unfortunately not. There is a `platform.php` key inside `composer.lock`, however this is often just a copy and paste of any `php` requirement set in the `require` section.

**Reply by rarkins** [reply]
It sounds like you'd need a new feature where Renovate would be instructed to select the minimum matching version instead of maximum.

**Reply by rarkins** [reply]
BTW Renovate itself may select versions to update composer.json to which you don't like, too.

**Reply by internalsystemerror** [reply]
But those would fail CI and open a PR giving us an indication as to why it failed. Our concern is primarily for the lock file maintenance. I will test adding a constraint and see how that changes things.

**Comment by viceice** [comment]
Looks like #13657 will improve this a lot
